Add automatic swagger 2.0 oauth2 token support on authenticated paths

borisirota opened the pull request:
add support for #42

borisirota commented:
@wcandillon I got stuck when found that the https://portal.28.io/api expects the token to be passed in the query and I still don't have my site running. Any ideas for sites like https://portal.28.io/api that expects the token to be passed in the header (Authorization: Bearer token) ? I didn't find any so far.

wcandillon commented:
The use case makes sense and would be useful for us at 28.io as well. I'm not sure about the implementation yet. I might take a look but I won't be able to do it in the short term.

borisirota commented:
@wcandillon I added support for sending token in a query and custom header. Do the w+test@28.io user has permissions to make some authenticated calls ? I added topic for that and tried some but received Unauthorized in all of them.

Anyway, I checked the code manually (no other choice meanwhile) and all the modes seems to be ok:
1. [default] Authorization Header = "Bearer token"
2. custom_name header = "token"
3. query param = "token"

wcandillon commented on lib/codegen.js:
```diff
@@ -36,6 +36,7 @@ var getViewForSwagger2 = function(opts, type){
     var data = {
         isNode: type === 'node',
         description: swagger.info.description,
+        isSecurity: swagger.securityDefinitions !== undefined,
```
`isSecure`

borisirota replied:
Done

wcandillon commented on tests/node-token.js:
```diff
@@ -0,0 +1,66 @@
+            var swagger = JSON.parse(fs.readFileSync('tests/apis/token.json', 'UTF-8'));
+            /*jshint evil:true*/
+            var Token = eval(CodeGen.getNodeCode({ className: 'Token', swagger: swagger }));
+            return new Token('https://portal.28.io/api');
```
@borisirota Can you rename this class to something more explicit? Like `ProtectedAPI`

wcandillon replied:
@borisirota This would be interesting to try with the default domain support at #55 (if we merge it first)

borisirota replied:
Done.
I added the host, basePath and schemas to the swagger spec (protected.json) so after the merge I can update the test to construct without the explicit url.

wcandillon commented on tests/node-token.js:
```diff
@@ -0,0 +1,66 @@
+vows.describe('Test Token').addBatch({
+    'Test Automatic Auth Token Assignment': {
+        topic: function(){
+            var swagger = JSON.parse(fs.readFileSync('tests/apis/token.json', 'UTF-8'));
```
@borisirota can you rename this file to something more explicit like `protected.json`

borisirota replied:
Done

wcandillon commented on tests/apis/token.json:
```diff
@@ -0,0 +1,155 @@
+                }
+            }
+        },
+        "/v1/_queries": {
```
@borisirota This endpoint won't work. Try `GET /project?token=...` and it should be a great test case.

borisirota replied:
Done

wcandillon commented:
@borisirota Great stuff! I commented on the diff.

borisirota commented:
@wcandillon update after code review + tests are passing :)

wcandillon commented on tests/node-protected.js:
```diff
@@ -0,0 +1,74 @@
+            },
+            'Calling operation with wrong token': {
+                topic: function (result) {
+                    result.protectedAPI.setToken(result.retval.body.access_token.slice(1), 'token', true);
```
@borisirota here do the write `setToken` invokation but overwrite the token using `getSecure({token: 'wrong'})`. That way to test that user can always overwrite the initial token.

borisirota replied:
Good point :+1: Done

wcandillon commented on templates/node-class.mustache:
```diff
@@ -14,6 +14,15 @@ var {{&className}} = (function(){
         this.domain = domain;
     }
 
+    {{#isSecure}}
```
@borisirota Do you think it could be possible to set the token in the constructor?

wcandillon replied:
@borisirota maybe these constructor now need to take an object of options as a parameter?
```javascript
new ProtectedAPI({
    domain: 'http://www.example.org/api',
    token: {
      value: ...
      ...
   }
});
```

borisirota replied:
@wcandillon Well, it can be as an option. We still need to ability to inject it because:
1. In the lifetime of the API class the token can be updated few times. For example: token's lifetime is a day and the browser tab is being open for a week -> it still should be possible to use the API.
2. One of the API's methods is GIVE_ME_THE_TOKEN so the token will be known after construction :)

What do you think ?

wcandillon replied:
@borisirota I agree completely. Just as an option ;-)

borisirota replied:
@wcandillon saw your second comment after posting mine. Do you want me to add this ability as an option?
It should support backwards compatibility? ->
```javascript
if (typeof domain === 'object') {
    ...
} else if (typeof domain === 'string') {
    ...
}
```
?

wcandillon replied:
@borisirota maybe it's an overkill idea, your call.

borisirota replied:
I like the options approach. Done

wcandillon commented on tests/node-protected.js:
```diff
@@ -0,0 +1,74 @@
+            var swagger = JSON.parse(fs.readFileSync('tests/apis/protected.json', 'UTF-8'));
+            /*jshint evil:true*/
+            var ProtectedAPI = eval(CodeGen.getNodeCode({ className: 'Protected', swagger: swagger }));
+            return new ProtectedAPI('https://portal.28.io/api');
```
@borisirota let's put `new ProtectedAPI()`

borisirota replied:
Done

wcandillon commented:
@borisirota Thanks a lot! I think we're very closed :+1: I added a couple more of minor comments, I hope you don't mind ;-) Thks again.

borisirota commented:
@wcandillon No worries :) I added support for all the discussed issues.

wcandillon commented on templates/angular-class.mustache:
```diff
@@ -11,13 +11,17 @@ angular.module('{{&moduleName}}', [])
+            cache = cache || ((typeof options === 'object') ? options.cache : cache);
             this.cache = cache;
+            {{#isSecure}}
+                this.token = (typeof options === 'object') ? options.token : {};
```
this.token default value needs to be either {} or undefined. In this line it seems that it can be both.

borisirota replied:
Ooppss :)
If the property exists it has to be {} because the setToken method try to access an object.

borisirota replied:
Done

wcandillon commented on templates/node-class.mustache:
```diff
@@ -10,13 +10,25 @@ var {{&className}} = (function(){
 
-    function {{&className}}(domain){
+    function {{&className}}(options){
+        var domain = (typeof options === 'object') ? options.domain : options;
```
same comment than https://github.com/wcandillon/swagger-js-codegen/pull/52/files#r24434064

borisirota replied:
Done

wcandillon commented:
@borisirota Thks a lot!

borisirota commented:
@wcandillon no problem :)
